api/filters/replacement/replacement.go

```diff
@@ -0,0 +1,136 @@
+	for _, t := range targets {
+		matches, err := getMatches(nodes, t)
+		if err != nil {
+			return nil, err
+		}
+		for _, n := range matches {
+			err := applyToNode(n, value, t)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+	return nodes, nil
+}
+
+func applyToNode(node *yaml.RNode, value *yaml.RNode, target *types.SelectedField) error {
+	if target.FieldPath != "" && len(target.FieldPaths) != 0 {
+		return fmt.Errorf("only one of fieldPath and fieldPaths is allowed for target")
+	}
+
+	var fieldPaths []string
+	switch {
+	case target.FieldPath == "" && len(target.FieldPaths) == 0:
+		fieldPaths = []string{DefaultFieldPath}
+	case target.FieldPath != "":
+		fieldPaths = []string{target.FieldPath}
+	default:
+		fieldPaths = target.FieldPaths
+	}
+
+	for i := range fieldPaths {
+		fieldPath := strings.Split(fieldPaths[i], ".")
+		// TODO (#3492): Support using map keys in the fieldPath (e.g. .spec.containers[name=nginx])
+		t, err := node.Pipe(yaml.Lookup(fieldPath...))
+		if err != nil {
+			return err
+		}
+		if t != nil {
+			// TODO (#3492): Use the `options` field to refine interpretation of the field
+			t.SetYNode(value.YNode())
+		}
+	}
+	return nil
+}
+
+func getReplacement(nodes []*yaml.RNode, r *types.Replacement) (*yaml.RNode, error) {
+	if r.SourceRaw == "" && r.SourceField == nil ||
+		r.SourceRaw != "" && r.SourceField != nil {
+		return nil, fmt.Errorf("exactly one of sourceField or sourceRaw must be present in replacement")
+	}
+
+	if r.SourceRaw != "" {
+		return yaml.Parse(r.SourceRaw)
+	}
+
+	matches, err := getMatches(nodes, r.SourceField)
+	if err != nil {
+		return nil, err
+	}
+	if len(matches) > 1 {
+		return nil, fmt.Errorf("found more than one resources matching from %v", r.SourceField)
+	}
+	if len(matches) == 0 {
+		return nil, fmt.Errorf("failed to find a resource matching from %v", r.SourceField)
+	}
+
+	fieldPath := r.SourceField.FieldPath
+	if fieldPath == "" {
+		fieldPath = DefaultFieldPath
+	}
+	fieldPathSlice := strings.Split(fieldPath, ".")
+
+	// TODO (#3492): Support using map keys in the fieldPath (e.g. .spec.containers[name=nginx])
+	rn, err := matches[0].Pipe(yaml.Lookup(fieldPathSlice...))
+	if err != nil {
+		return nil, err
+	}
+	return rn, nil
+}
+
+// getMatches finds the node that matches the SelectedField
+// TODO (#3492): Don't include matches listed in the `deny` field
+func getMatches(nodes []*yaml.RNode, objRef *types.SelectedField) ([]*yaml.RNode, error) {
+	var matches []*yaml.RNode
+	allow := objRef.Allow
+	for _, n := range nodes {
+		ns, err := n.GetNamespace()
+		if err != nil {
+			return nil, err
+		}
+		apiVersion := yaml.GetValue(n.Field(yaml.APIVersionField).Value)
+		if (n.GetName() == allow.Name || allow.Name == "") &&
+			(n.GetKind() == allow.Kind || allow.Kind == "") &&
+			(ns == allow.Namespace || allow.Namespace == "") &&
+			(apiVersion == allow.ApiVersion() || allow.ApiVersion() == "") {
```

Comment:
please put this boolean expression in a bool function named `allowed` (or something :-)